File: qh3server/qh3serverplugin_ts_proj/src/userserver/userserver.ts
import * as ffi from 'ffi-napi';
import { qh3serversdk } from '../helpers/qh3serversdk';
import * as ref from 'ref-napi';
import { essentials } from '../helpers/essentials';
import { qmongo } from '../helpers/qmongo';
import qhiredis from '../helpers/qhiredis';
import qzookeeper from '../helpers/qzookeeper';
import { debug_print, debug_error, LOG_LEVEL_0, LOG_LEVEL_4 } from '../helpers/sdktypes';
import { serverconfig } from '../helpers/serverconfig';
import * as path from 'path';

export namespace server {
    type type_api_callback = (user_server_interface: interface_userserver, api_instance: interface_api, path: string, buffer: string, len: number) => Promise<string | null | any>;

    export interface interface_userserver {
        get_mongo_driver(): qmongo | null;
        get_hiredis_driver(): qhiredis | null;
        get_qzookeeper_driver(): qzookeeper | null;
        get_zkconfig(): serverconfig | null;
    }

    export interface interface_api {
        get_path(): string;
        get_post_cb(): type_api_callback;
    }

    export class userserver implements interface_userserver {
        private static __LOGTAG__: string = `userserver`;
        private router_config : qh3serversdk.qh3_router_input_config = {
            router_address: `127.0.0.1:4004`,
            mongodb_uri: `mongodb://3.109.144.159:27017`,
            redis_address: `3.109.144.159:6379`,
            zk_uri: `3.109.144.159:2181`,
            root_dir: process.cwd(),
            command_port: 4010,
            router_port_return: 4005,
            app_id: `qh3serverplugin-app`
        };
        private mongo : qmongo | null = null;
        private hiredis: qhiredis | null = null;
        private qzk: qzookeeper | null = null;
        private zkconfig: serverconfig | null = null;
        private api_callbacks: Map<string, interface_api> = new Map();

        public get_mongo_driver(): qmongo | null {
            return this.mongo;
        }
        public get_hiredis_driver(): qhiredis | null {
            return this.hiredis;
        }
        public get_qzookeeper_driver(): qzookeeper | null {
            return this.qzk;
        }
        public get_zkconfig(): serverconfig | null {
            return this.zkconfig;
        }

        protected on_server_pre_start = ffi.Callback('void', ['pointer'], (native_server: Buffer) => {
            // debug_print(LOG_LEVEL_4, userserver.__LOGTAG__, `on_server_pre_start`);
        }) as unknown as qh3serversdk.type_on_server_pre_start;
        protected on_server_start = ffi.Callback('void', ['pointer'], async (native_server: Buffer) => {
            // debug_print(LOG_LEVEL_4, userserver.__LOGTAG__, `on_server_start`);
        }) as unknown as qh3serversdk.type_on_server_start;
        protected on_server_stop = ffi.Callback('void', ['pointer'], (native_server: Buffer) => {
            // debug_print(LOG_LEVEL_4, userserver.__LOGTAG__, `on_server_stop:`);
        }) as unknown as qh3serversdk.type_on_server_stop;
        protected on_server_error = ffi.Callback('void', ['pointer'], (native_server: Buffer, error_code: number) => {
            // debug_error(userserver.__LOGTAG__, `on_server_error: ${error_code}`);
        }) as unknown as qh3serversdk.type_on_server_error;
        protected on_server_parse = ffi.Callback('void', ['pointer', 'pointer', 'string', 'string', 'int'], async (native_server: Buffer, conn: Buffer, path: string, buffer: string, len: number) => {
            debug_print(LOG_LEVEL_4, userserver.__LOGTAG__, `on_server_parse: ${path}, ${buffer}, len ${len}`);
            let result: string | null = null;

            if (this.api_callbacks.has(path)) {
                let api_instance: interface_api | any= this.api_callbacks.get(path);
                result = await api_instance?.get_post_cb()?.(this, api_instance, path, buffer, len);
            }

            if (result) {
                qh3serversdk.qh3serverplugin.qh3server_try_send_response(native_server, conn, result, result.length, null, 0);
            } else {
                qh3serversdk.qh3serverplugin.qh3server_try_send_response(native_server, conn, `{}`, 2, null, 0);
            }

            // const response_string = `{msg:\"test message\"}`;
            // // Use strdup to allocate and return a copy of the string
            // result = qh3serversdk.libc.strdup(response_string);
            // if (result.isNull()) {
            //     throw new Error('Memory allocation failed using strdup.');
            // }
            // return result; // Return the pointer to C++
        }) as unknown as qh3serversdk.type_on_server_parse;

        public register_api(api_instance : interface_api) : void {
            if (this.api_callbacks.has(api_instance.get_path())) {
                return;
            }
            this.api_callbacks.set(api_instance.get_path(), api_instance);
            debug_print(LOG_LEVEL_0, userserver.__LOGTAG__, `api registered - ${api_instance.get_path()}`);
        }

        public unregister_api(path:string) : void {
            if (this.api_callbacks.has(path)) {
                this.api_callbacks.delete(path);
                debug_print(LOG_LEVEL_0, userserver.__LOGTAG__, `api un-registered - ${path}`);
            }
        }

        public unregister_api_instance(api_instance : interface_api) : void {
            if (this.api_callbacks.has(api_instance.get_path())) {
                this.api_callbacks.delete(api_instance.get_path());
                debug_print(LOG_LEVEL_0, userserver.__LOGTAG__, `api un-registered - ${api_instance.get_path()}`);
            }
        }

        public async run(native_router: Buffer) : Promise<void> {
            // mongo setup
            this.mongo = new qmongo("", "gsdk_mongodb", this.router_config.mongodb_uri);
            await this.mongo?.connect();

            // redis setup
            const redis_ip_port = essentials.extract_ip_and_port(this.router_config.redis_address);
            if (redis_ip_port) {
                const [ip, port]: [string, number] = redis_ip_port;
                this.hiredis = new qhiredis("hiredis", ip, port, "gsdkuser", "Fr0gmoon123");
                await this.hiredis.connect_redis();
            } else {
                debug_error(userserver.__LOGTAG__, "Invalid hiredis address format");
                return;
            }

            // zookeeper setup
            const zk_ip_port = essentials.extract_ip_and_port(this.router_config.zk_uri);
            if (zk_ip_port) {
                const [ip, port]: [string, number] = zk_ip_port;
                this.qzk = new qzookeeper(this.router_config.zk_uri);
                await this.qzk.connect();
            } else {
                debug_error(userserver.__LOGTAG__, "Invalid zk address format");
                return;
            }

            // load server config
            this.zkconfig = new serverconfig(this.qzk, null);
            const config_path = path.join(this.router_config.root_dir, 'configs', 'dev', 'runtime-config.json');
            debug_print(LOG_LEVEL_4, userserver.__LOGTAG__, `reading ${config_path}`);
            await this.zkconfig.load(config_path, this.qzk, `/qh3server`);

            // server spawn
            qh3serversdk.qh3serverplugin.spawn_qh3server(
                native_router,
                this.router_config.router_address,
                this.router_config.mongodb_uri,
                this.router_config.redis_address,
                this.router_config.zk_uri,
                this.router_config.root_dir,
                this.router_config.command_port,
                this.router_config.router_port_return,
                this.router_config.app_id,
                this.on_server_pre_start,
                this.on_server_start,
                this.on_server_stop,
                this.on_server_error,
                this.on_server_parse
            );
        }
    }
}

export default server;

File: serverplugin_lua_proj/src/userserver/userserver.lua
local serversdk = require("src.helpers.serversdk")
local sdktypes = require("src.helpers.sdktypes")
local routerserver = require("src.userserver.router")
local qmongo = require("src.helpers.qmongo")
local qhiredis = require("src.helpers.qhiredis")
local qzookeeper = require("src.helpers.qzookeeper")
local essentials = require("src.helpers.essentials")
local serverconfig = require("src.helpers.serverconfig")
local ffi = require("ffi")

ffi.cdef[[
    typedef struct qh3server qh3server;
    typedef void (*type_on_server_pre_start)(qh3server* server, void* user_arg);
    typedef void (*type_on_server_start)(qh3server* server, void* user_arg, const char* ip, uint16_t port);
    typedef void (*type_on_server_stop)(qh3server* server, void* user_arg);
    typedef void (*type_on_server_error)(qh3server* server, void* user_arg, int error_code);
    typedef void (*type_on_server_parse)(
        void* server, 
        void* user_arg,
        uint8_t* cid,
        uint16_t cid_len,
        const char* path,
        const char* buffer,
        unsigned long len,
        const char* headers_buffer,
        unsigned long headers_buffer_size
    );
]]

local server = {}
server.userserver = {}
server.userserver.__LOGTAG__ = "userserver"

-- Constructor
function server.userserver:new()
    local obj = {}
    obj.api_callbacks = {}
    obj.mongo = nil
    obj.hiredis = nil
    obj.zk = nil
    obj.zkconfig = nil
    obj.zk_value_change_callback_handle = nil
    obj.request_counter = 0
    obj.total_execution_time = 0
    obj.start_time = serversdk.sdklib.get_current_time_in_ms()
    setmetatable(obj, self)
    self.registry_id = serversdk.serverplugin.add_lua_object_to_registry(self)
    self.__index = self
    return obj
end

function server.userserver:get_mongo_driver()
    return self.mongo
end

function server.userserver:get_hiredis_driver()
    return self.hiredis
end

function server.userserver:get_qzookeeper_driver()
    return self.zk
end

function server.userserver:get_zkconfig()
    return self.zkconfig
end

function server.userserver:register_api(api_instance)
    if self.api_callbacks[api_instance:get_path()] then
        return
    end
    self.api_callbacks[api_instance:get_path()] = api_instance
    sdktypes.debug_print(sdktypes.LOG_LEVEL_0, "userserver", "api registered - " .. api_instance:get_path())
end

function server.userserver:unregister_api(path)
    if self.api_callbacks[path] then
        self.api_callbacks[path] = nil
        sdktypes.debug_print(sdktypes.LOG_LEVEL_0, "userserver", "api un-registered - " .. path)
    end
end

function server.userserver:unregister_api_instance(api_instance)
    if self.api_callbacks[api_instance:get_path()] then
        self.api_callbacks[api_instance:get_path()] = nil
        sdktypes.debug_print(sdktypes.LOG_LEVEL_0, "userserver", "api un-registered - " .. api_instance:get_path())
    end
end

local function on_zk_value_change(path, data, context)
    -- sdktypes.debug_print(sdktypes.LOG_LEVEL_0, "userserver", "api un-registered - " .. api_instance:get_path())
    print(string.format("Value changed! Path: %s, Data: %s", ffi.string(path), ffi.string(data)))
end

-- Assign FFI casted callback functions
server.userserver.on_server_pre_start = ffi.cast("type_on_server_pre_start", function(native_server, user_arg)
    print("on_server_pre_start triggered")
end)

server.userserver.on_server_start = ffi.cast("type_on_server_start", function(native_server, user_arg, ip, port)
    print("on_server_start triggered: IP = " .. ffi.string(ip) .. ", Port = " .. port)
    local hash_key = "servers:" .. serversdk.serverplugin.get_device_public_ip()
    local field_key = "server-" .. port
    local field_value = ffi.string(ip) .. ":" .. port
    local thiz = serversdk.serverplugin.get_from_registry(user_arg);
    thiz.hiredis:set_hash_value(hash_key, field_key, field_value)
end)

server.userserver.on_server_stop = ffi.cast("type_on_server_stop", function(native_server, user_arg)
    print("on_server_stop triggered")
end)

server.userserver.on_server_error = ffi.cast("type_on_server_error", function(native_server, user_arg, error_code)
    print("on_server_error triggered: Error Code = " .. error_code)
end)

server.userserver.on_server_parse = ffi.cast("type_on_server_parse", function(
    native_server, native_user_arg, native_cid, cid_len, native_path, native_buffer, len, native_headers_buffer, native_headers_buffer_size)
    local thiz = serversdk.serverplugin.get_from_registry(native_user_arg);
    thiz.request_counter = thiz.request_counter + 1
    local parse_start_time = serversdk.sdklib.get_current_time_in_ms();

    -- Using pcall to safely call the process_request function
    local success, result = pcall(function()
        return thiz.process_request(native_server, native_user_arg, native_cid, cid_len, native_path, native_buffer, len, native_headers_buffer, native_headers_buffer_size)
    end)

    -- Check if the call was successful
    if success then
        if result then
            serversdk.serverplugin.qh3server_try_send_response(native_server, native_cid, cid_len, result, #result, nil, 0)
        else
            serversdk.serverplugin.qh3server_try_send_response(native_server, native_cid, cid_len, "{}", 2, nil, 0)
        end
    else
        local error_msg = result  -- `result` will contain the error message
        print("Error occurred during process_request: " .. error_msg)
        serversdk.serverplugin.qh3server_try_send_response(native_server, native_cid, cid_len, "{}", 2, nil, 0)
    end

    local execution_time = serversdk.sdklib.get_current_time_in_ms() - parse_start_time
    thiz.total_execution_time = thiz.total_execution_time + execution_time
    thiz:calculate_rps()
end)

server.userserver.process_request = ffi.cast("type_on_server_parse", function(
    native_server, native_user_arg, native_cid, cid_len, native_path, native_buffer, len, native_headers_buffer, native_headers_buffer_size)
    -- local parse_start_time = serversdk.sdklib.get_current_time_in_ms();
    local thiz = serversdk.serverplugin.get_from_registry(native_user_arg);
    local lua_path = ffi.string(native_path)
    -- print("on_server_parse triggered: Path = " .. lua_path)
    local result = "{}"
    if thiz.api_callbacks and thiz.api_callbacks[lua_path] then
        local api_instance = thiz.api_callbacks[lua_path]
        if api_instance and api_instance.get_post_cb then
            local post_cb = api_instance:get_post_cb()
            if post_cb then
                local c_result = post_cb(native_server, native_user_arg, native_cid, cid_len, native_path, native_buffer, len, native_headers_buffer, native_headers_buffer_size)
                result = ffi.string(c_result)
            end
        end
    end
    return result;
end)

-- Function to calculate and display requests per second on the same line
function server.userserver:calculate_rps()
    local current_time = serversdk.sdklib.get_current_time_in_ms()
    local elapsed_time_ms = current_time - self.start_time

    if elapsed_time_ms >= 1000 then
        local rps = (self.request_counter * 1000) / elapsed_time_ms
        local avg_execution_time = self.total_execution_time / math.max(self.request_counter, 1) -- Avoid division by zero

        io.write(string.format("\rRequests per second: %.2f | Avg execution time: %.2f ms", tonumber(rps), tonumber(avg_execution_time)))
        -- io.write(string.format("\rRequests per second: %.2f", tonumber(rps)))
        io.flush() -- Ensure the output is written to the terminal

        -- Reset the counter and time for the next interval
        self.request_counter = 0
        self.total_execution_time = 0
        self.start_time = current_time
    end
end

-- Run function to start the router
function server.userserver:run(native_router)
    -- Create a qmongo instance
    self.mongo = qmongo:new("", "gsdk_mongodb", routerserver.router.router_config.mongodb_uri, function(error)
        print("mongo error:", error)
    end)
    self.mongo:connect()
    local redis_ip, redis_port = essentials.extract_ip_and_port(routerserver.router.router_config.redis_address)
    if not redis_ip or not redis_port then
        print("Invalid redis IP and port format")
    end

    self.hiredis = qhiredis:new("hiredis", redis_ip, redis_port, "gsdkuser", "Fr0gmoon123")
    if self.hiredis ~= nil and self.hiredis:connect(0) ~= sdktypes.EXIT_SUCCESS then
        sdktypes.debug_print(sdktypes.LOG_LEVEL_0, server.userserver.__LOGTAG__, "hiredis connect failed !!!");
        return;
    end
    
    self.zk = qzookeeper:new("lua-zk")
    if self.zk ~= nil and self.zk:connect(routerserver.router.router_config.zk_uri) ~= sdktypes.EXIT_SUCCESS then
        sdktypes.debug_print(sdktypes.LOG_LEVEL_0, server.userserver.__LOGTAG__, "zk connect failed !!!");
        return;
    else
        self.zk_value_change_callback_handle = self.zk:register_value_change_callback(on_zk_value_change, self.registry_id)
    end

    -- load server config
    self.zkconfig = serverconfig.serverconfig:new(self.zk, nil)
    local release_type = (os.getenv("SERVER_ENV") == "production") and 'prod' or 'dev'
    local config_path = routerserver.router.router_config.root_dir .. '/configs/' .. release_type .. '/runtime-config.json'
    sdktypes.debug_print(sdktypes.LOG_LEVEL_4, server.userserver.__LOGTAG__, "reading " .. config_path)
    self.zkconfig:load(config_path, self.zk, "/qh3server")

    -- local temp = self.zk:get_data("/qh3server/server_config/user_token_expiry_time", "0");

    serversdk.serverplugin.spawn_qh3server(
        native_router,
        routerserver.router.router_config.router_address, 
        routerserver.router.router_config.mongodb_uri, 
        routerserver.router.router_config.redis_address, 
        routerserver.router.router_config.zk_uri, 
        routerserver.router.router_config.root_dir, 
        routerserver.router.router_config.command_port,
        routerserver.router.router_config.router_port_return,
        routerserver.router.router_config.app_id,
        self.on_server_pre_start,
        self.on_server_start,
        self.on_server_stop,
        self.on_server_error,
        self.on_server_parse,
        self
    )
end

-- Simulate destructor using the __gc metamethod
function server.userserver:__gc()
    print(self.name .. " object is being garbage collected.")
end

return server
